File: examples/python/soundcloud.py
"""
Cognima API - SoundCloud Examples (Python)

This file demonstrates how to use the SoundCloud endpoints
to search and download tracks.
"""

import requests
import time
from typing import List, Dict

# API Configuration
BASE_URL = 'https://cog.api.br/api/v1'
API_KEY = 'YOUR_API_KEY_HERE'

# Configure headers
headers = {
    'apikey': API_KEY
}

def search_soundcloud(query: str, limit: int = 10) -> List[Dict]:
    """
    Search for tracks on SoundCloud
    
    Args:
        query: Track name or artist
        limit: Number of results
    
    Returns:
        List of tracks
    """
    print(f'\n=== Searching SoundCloud: "{query}" ===\n')
    
    params = {
        'q': query,
        'limit': limit
    }
    
    response = requests.get(f'{BASE_URL}/soundcloud/search', params=params, headers=headers)
    response.raise_for_status()
    
    data = response.json()
    results = data['results']
    
    print(f"Found {data['total']} results:\n")
    
    for i, track in enumerate(results, 1):
        print(f"{i}. {track['title']}")
        print(f"   Artist ID: {track['artist']}")
        print(f"   Duration: {track['duration']}s")
        print(f"   Plays: {track['playback_count']:,}")
        print(f"   Likes: {track['likes_count']:,}")
        print(f"   URL: {track['permalink_url']}")
        print()
    
    return results

def search_one_track(query: str) -> Dict:
    """
    Search for a specific track (returns only the first result)
    
    Args:
        query: Track name or artist
    
    Returns:
        First track found
    """
    print(f'\n=== Searching One Track: "{query}" ===\n')
    
    params = {'q': query}
    
    response = requests.get(f'{BASE_URL}/soundcloud/search-one', params=params, headers=headers)
    response.raise_for_status()
    
    data = response.json()
    track = data['result']
    
    print('✅ Track found:\n')
    print(f"Title: {track['title']}")
    print(f"Artist ID: {track['artist']}")
    print(f"Duration: {track['duration']}s")
    print(f"Plays: {track['playback_count']:,}")
    print(f"Likes: {track['likes_count']:,}")
    print(f"Genre: {track['genre']}")
    print(f"URL: {track['permalink_url']}")
    
    return track

def search_multiple_tracks(queries: List[str]) -> List[Dict]:
    """
    Search for multiple different tracks
    
    Args:
        queries: List of track names
    
    Returns:
        List of tracks found
    """
    print('\n=== Searching Multiple Tracks ===\n')
    
    all_results = []
    
    for query in queries:
        params = {'q': query}
        response = requests.get(f'{BASE_URL}/soundcloud/search-one', params=params, headers=headers)
        response.raise_for_status()
        
        result = response.json()['result']
        all_results.append(result)
        print(f"✅ Found: {result['title']}")
    
    print(f'\n📋 Total tracks found: {len(all_results)}')
    return all_results

def get_recommendations(query: str, count: int = 5) -> List[Dict]:
    """
    Get track recommendations based on a search
    
    Args:
        query: Track name or artist
        count: Number of recommendations
    
    Returns:
        List of recommended tracks
    """
    print(f'\n=== Getting {count} Recommendations for: "{query}" ===\n')
    
    params = {
        'q': query,
        'limit': count
    }
    
    response = requests.get(f'{BASE_URL}/soundcloud/search', params=params, headers=headers)
    response.raise_for_status()
    
    results = response.json()['results']
    
    print('🎵 Recommendations:\n')
    for i, track in enumerate(results, 1):
        print(f"{i}. {track['title']}")
        print(f"   Plays: {track['playback_count']:,} | Likes: {track['likes_count']:,}")
    
    return results

def download_track(url: str) -> Dict:
    """
    Download track from SoundCloud by URL
    
    Args:
        url: SoundCloud track URL
    
    Returns:
        Dictionary with download information
    """
    print('\n=== Downloading Track from SoundCloud ===\n')
    print(f'URL: {url}\n')
    
    params = {'url': url}
    
    response = requests.get(f'{BASE_URL}/soundcloud/download', params=params, headers=headers)
    response.raise_for_status()
    
    data = response.json()
    
    if data['success']:
        download = data['data']
        print('✅ Download Ready!\n')
        print(f"Title: {download['title']}")
        print(f"Artist: {download['artist']}")
        print(f"Thumbnail: {download['thumbnail']}")
        print(f"\n📥 Download URL: {download['downloadUrl']}")
        
        return download
    
    return {}

def search_and_download(query: str) -> Dict:
    """
    Search and download track automatically
    
    Args:
        query: Track name or artist
    
    Returns:
        Dictionary with track and download information
    """
    print(f'\n=== Searching and Downloading: "{query}" ===\n')
    
    params = {'q': query}
    
    response = requests.get(f'{BASE_URL}/soundcloud/search-download', params=params, headers=headers)
    response.raise_for_status()
    
    data = response.json()
    
    if data['success']:
        track = data['track']
        download = data['download']
        
        print('✅ Track Found and Download Ready!\n')
        print('🎵 Track Info:')
        print(f"   Title: {track['title']}")
        print(f"   Artist ID: {track['artist']}")
        print(f"   Duration: {track['duration']}s")
        print(f"   Plays: {track['playback_count']:,}")
        print(f"   Likes: {track['likes_count']:,}")
        print(f"   URL: {track['permalink_url']}")
        print('\n📥 Download Info:')
        print(f"   Title: {download['title']}")
        print(f"   Artist: {download['artist']}")
        print(f"   Thumbnail: {download['thumbnail']}")
        print(f"\n📥 Download URL: {download['downloadUrl']}")
        
        return {'track': track, 'download': download}
    
    return {}

def search_by_genre(genre: str, limit: int = 10) -> List[Dict]:
    """
    Search by genre and get top tracks
    
    Args:
        genre: Genre name
        limit: Number of results
    
    Returns:
        List of tracks from the genre
    """
    print(f'\n=== Searching {genre} tracks ===\n')
    
    params = {
        'q': genre,
        'limit': limit
    }
    
    response = requests.get(f'{BASE_URL}/soundcloud/search', params=params, headers=headers)
    response.raise_for_status()
    
    results = response.json()['results']
    
    print(f'🎵 Top {genre} Tracks:\n')
    for i, track in enumerate(results, 1):
        print(f"{i}. {track['title']}")
        print(f"   Genre: {track['genre']}")
        print(f"   Plays: {track['playback_count']:,}")
    
    return results

def get_trending_tracks(query: str, limit: int = 10) -> List[Dict]:
    """
    Get trending tracks based on play count
    
    Args:
        query: Search query
        limit: Number of results
    
    Returns:
        List of trending tracks
    """
    print(f'\n=== Getting Trending Tracks: "{query}" ===\n')
    
    params = {
        'q': query,
        'limit': limit
    }
    
    response = requests.get(f'{BASE_URL}/soundcloud/search', params=params, headers=headers)
    response.raise_for_status()
    
    results = response.json()['results']
    
    # Sort by playback count
    trending = sorted(results, key=lambda x: x['playback_count'], reverse=True)
    
    print('🔥 Trending Tracks:\n')
    for i, track in enumerate(trending, 1):
        print(f"{i}. {track['title']}")
        print(f"   Plays: {track['playback_count']:,}")
        print(f"   Likes: {track['likes_count']:,}")
    
    return trending

def download_multiple_tracks(queries: List[str]) -> List[Dict]:
    """
    Search and download multiple tracks
    
    Args:
        queries: List of track searches
    
    Returns:
        List of download results
    """
    print('\n=== Downloading Multiple Tracks ===\n')
    
    downloads = []
    
    for query in queries:
        try:
            result = search_and_download(query)
            downloads.append(result)
            print(f"✅ Downloaded: {result['download']['title']}")
            time.sleep(1)  # Delay to avoid rate limiting
        except Exception as e:
            print(f"❌ Failed to download: {query} - {str(e)}")
        
        print()
    
    print(f'\n📋 Downloaded {len(downloads)} tracks')
    return downloads

def create_playlist_from_searches(queries: List[str]) -> List[Dict]:
    """
    Create a playlist by searching for multiple tracks
    
    Args:
        queries: List of track searches
    
    Returns:
        List of tracks for the playlist
    """
    print('\n=== Creating Playlist ===\n')
    
    playlist = []
    
    for i, query in enumerate(queries, 1):
        print(f'Adding track {i}/{len(queries)}...')
        
        try:
            track = search_one_track(query)
            playlist.append(track)
            print(f'✅ Added: {track["title"]}\n')
        except Exception as e:
            print(f'❌ Failed to add: {query} - {str(e)}\n')
    
    print(f'\n📋 Playlist created with {len(playlist)} tracks')
    return playlist

# ===================
# EXAMPLES
# ===================

def main():
    try:
        # Example 1: Search for multiple tracks
        search_soundcloud('te vi de canto', 5)
        
        # Example 2: Search for a specific track
        search_one_track('te vi dançando')
        
        # Example 3: Search multiple different tracks
        queries = [
            'lofi hip hop',
            'deep house mix',
            'ambient music'
        ]
        search_multiple_tracks(queries)
        
        # Example 4: Get recommendations
        get_recommendations('electronic music', 5)
        
        # Example 5: Download track by URL
        download_track('https://soundcloud.com/jose-luiiz-ii/ro-rosa-te-vi-de-canto-prod')
        
        # Example 6: Search and download automatically (recommended!)
        search_and_download('te vi de canto')
        
        # Example 7: Search by genre
        search_by_genre('hip hop', 5)
        
        # Example 8: Get trending tracks
        get_trending_tracks('remix', 5)
        
        # Example 9: Create a playlist
        playlist_queries = [
            'lofi beats',
            'chill vibes',
            'study music'
        ]
        create_playlist_from_searches(playlist_queries)
        
        # Example 10: Download multiple tracks
        download_queries = [
            'te vi dançando',
            'electronic music',
            'hip hop beat'
        ]
        download_multiple_tracks(download_queries)
        
    except Exception as e:
        print(f'Error: {str(e)}')
        exit(1)

if __name__ == '__main__':
    main()

File: examples/nodejs/soundcloud.js
/**
 * Cognima API - SoundCloud Examples (Node.js)
 * 
 * This file demonstrates how to use the SoundCloud endpoints
 * to search and download tracks.
 */

const axios = require('axios');

// API Configuration
const BASE_URL = 'https://cog.api.br/api/v1';
const API_KEY = 'YOUR_API_KEY_HERE';

// Configure axios with API key
const api = axios.create({
  baseURL: BASE_URL,
  headers: {
    'apikey': API_KEY
  }
});

/**
 * Search for tracks on SoundCloud
 * @param {string} query - Track name or artist
 * @param {number} limit - Number of results
 */
async function searchSoundCloud(query, limit = 10) {
  try {
    console.log(`\n=== Searching SoundCloud: "${query}" ===\n`);
    
    const response = await api.get('/soundcloud/search', {
      params: { 
        q: query, 
        limit 
      }
    });
    
    const { total, results } = response.data;
    
    console.log(`Found ${total} results:\n`);
    
    results.forEach((track, index) => {
      console.log(`${index + 1}. ${track.title}`);
      console.log(`   Artist ID: ${track.artist}`);
      console.log(`   Duration: ${track.duration}s`);
      console.log(`   Plays: ${track.playback_count.toLocaleString()}`);
      console.log(`   Likes: ${track.likes_count.toLocaleString()}`);
      console.log(`   URL: ${track.permalink_url}`);
      console.log();
    });
    
    return results;
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Search for a specific track
 * @param {string} query - Track name or artist
 */
async function searchOneTrack(query) {
  try {
    console.log(`\n=== Searching One Track: "${query}" ===\n`);
    
    const response = await api.get('/soundcloud/search-one', {
      params: { q: query }
    });
    
    const { result } = response.data;
    
    console.log('✅ Track found:\n');
    console.log(`Title: ${result.title}`);
    console.log(`Artist ID: ${result.artist}`);
    console.log(`Duration: ${result.duration}s`);
    console.log(`Plays: ${result.playback_count.toLocaleString()}`);
    console.log(`Likes: ${result.likes_count.toLocaleString()}`);
    console.log(`Genre: ${result.genre}`);
    console.log(`URL: ${result.permalink_url}`);
    
    return result;
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Search for multiple different tracks
 * @param {Array<string>} queries - Array of track names
 */
async function searchMultipleTracks(queries) {
  try {
    console.log('\n=== Searching Multiple Tracks ===\n');
    
    const allResults = [];
    
    for (const query of queries) {
      const response = await api.get('/soundcloud/search-one', {
        params: { q: query }
      });
      
      allResults.push(response.data.result);
      console.log(`✅ Found: ${response.data.result.title}`);
    }
    
    console.log(`\n📋 Total tracks found: ${allResults.length}`);
    return allResults;
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Get track recommendations based on a search
 * @param {string} query - Track name or artist
 * @param {number} count - Number of recommendations
 */
async function getRecommendations(query, count = 5) {
  try {
    console.log(`\n=== Getting ${count} Recommendations for: "${query}" ===\n`);
    
    const response = await api.get('/soundcloud/search', {
      params: { 
        q: query, 
        limit: count 
      }
    });
    
    const { results } = response.data;
    
    console.log('🎵 Recommendations:\n');
    results.forEach((track, index) => {
      console.log(`${index + 1}. ${track.title}`);
      console.log(`   Plays: ${track.playback_count.toLocaleString()} | Likes: ${track.likes_count.toLocaleString()}`);
    });
    
    return results;
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Download track from SoundCloud by URL
 * @param {string} url - SoundCloud track URL
 */
async function downloadTrack(url) {
  try {
    console.log(`\n=== Downloading Track from SoundCloud ===\n`);
    console.log(`URL: ${url}\n`);
    
    const response = await api.get('/soundcloud/download', {
      params: { url }
    });
    
    const { success, data } = response.data;
    
    if (success) {
      console.log('✅ Download Ready!\n');
      console.log(`Title: ${data.title}`);
      console.log(`Artist: ${data.artist}`);
      console.log(`Thumbnail: ${data.thumbnail}`);
      console.log(`\n📥 Download URL: ${data.downloadUrl}`);
      
      return data;
    }
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Search and download track automatically
 * @param {string} query - Track name or artist
 */
async function searchAndDownload(query) {
  try {
    console.log(`\n=== Searching and Downloading: "${query}" ===\n`);
    
    const response = await api.get('/soundcloud/search-download', {
      params: { q: query }
    });
    
    const { success, track, download } = response.data;
    
    if (success) {
      console.log('✅ Track Found and Download Ready!\n');
      console.log('🎵 Track Info:');
      console.log(`   Title: ${track.title}`);
      console.log(`   Artist ID: ${track.artist}`);
      console.log(`   Duration: ${track.duration}s`);
      console.log(`   Plays: ${track.playback_count.toLocaleString()}`);
      console.log(`   Likes: ${track.likes_count.toLocaleString()}`);
      console.log(`   URL: ${track.permalink_url}`);
      console.log('\n📥 Download Info:');
      console.log(`   Title: ${download.title}`);
      console.log(`   Artist: ${download.artist}`);
      console.log(`   Thumbnail: ${download.thumbnail}`);
      console.log(`\n📥 Download URL: ${download.downloadUrl}`);
      
      return { track, download };
    }
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Search by genre and get top tracks
 * @param {string} genre - Genre name
 * @param {number} limit - Number of results
 */
async function searchByGenre(genre, limit = 10) {
  try {
    console.log(`\n=== Searching ${genre} tracks ===\n`);
    
    const response = await api.get('/soundcloud/search', {
      params: { 
        q: genre, 
        limit 
      }
    });
    
    const { results } = response.data;
    
    console.log(`🎵 Top ${genre} Tracks:\n`);
    results.forEach((track, index) => {
      console.log(`${index + 1}. ${track.title}`);
      console.log(`   Genre: ${track.genre}`);
      console.log(`   Plays: ${track.playback_count.toLocaleString()}`);
    });
    
    return results;
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Get trending tracks based on play count
 * @param {string} query - Search query
 * @param {number} limit - Number of results
 */
async function getTrendingTracks(query, limit = 10) {
  try {
    console.log(`\n=== Getting Trending Tracks: "${query}" ===\n`);
    
    const response = await api.get('/soundcloud/search', {
      params: { 
        q: query, 
        limit 
      }
    });
    
    const { results } = response.data;
    
    // Sort by playback count
    const trending = results.sort((a, b) => b.playback_count - a.playback_count);
    
    console.log('🔥 Trending Tracks:\n');
    trending.forEach((track, index) => {
      console.log(`${index + 1}. ${track.title}`);
      console.log(`   Plays: ${track.playback_count.toLocaleString()}`);
      console.log(`   Likes: ${track.likes_count.toLocaleString()}`);
    });
    
    return trending;
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

// ===================
// EXAMPLES
// ===================

async function main() {
  try {
    // Example 1: Search for multiple tracks
    await searchSoundCloud('te vi de canto', 5);
    
    // Example 2: Search for a specific track
    await searchOneTrack('te vi dançando');
    
    // Example 3: Search multiple different tracks
    const queries = [
      'lofi hip hop',
      'deep house mix',
      'ambient music'
    ];
    await searchMultipleTracks(queries);
    
    // Example 4: Get recommendations
    await getRecommendations('electronic music', 5);
    
    // Example 5: Download track by URL
    await downloadTrack('https://soundcloud.com/jose-luiiz-ii/ro-rosa-te-vi-de-canto-prod');
    
    // Example 6: Search and download automatically (recommended!)
    await searchAndDownload('te vi de canto');
    
    // Example 7: Search by genre
    await searchByGenre('hip hop', 5);
    
    // Example 8: Get trending tracks
    await getTrendingTracks('remix', 5);
    
  } catch (error) {
    console.error('Main error:', error.message);
    process.exit(1);
  }
}

// Run examples
if (require.main === module) {
  main();
}

module.exports = {
  searchSoundCloud,
  searchOneTrack,
  searchMultipleTracks,
  getRecommendations,
  downloadTrack,
  searchAndDownload,
  searchByGenre,
  getTrendingTracks
};
